docs/ReferenceGuides/UnderscoredAttributes.md

```diff
@@ -0,0 +1,328 @@
+
+Most notably, default argument expressions are implicitly
+`@_alwaysEmitIntoClient`, which means that adding
+a default argument to a function which did not have one previously
+does not break ABI.
+
+## `@_borrowed`
+
+## `@_cdecl`
+
+Similar to `@_silgen_name` but uses the C calling convention.
+This attribute doesn't have very well-defined semantics.
+
+Type bridging is not done, so the parameter and return types should
+correspond directly to types accessible in C.
+In most cases, it is preferable to define a static method on an `@objc`
+class instead of using `@_cdecl`.
+
+For potential ideas on stabilization, see
+[Formalizing `@cdecl`](https://forums.swift.org/t/formalizing-cdecl/40677).
+
+## `@_disfavoredOverload`
+
+Marks a declaration that the type checker should try to avoid using. When the
+expression type checker is considering overloads, it will prefer a solution
+with fewer `@_disfavoredOverload` declarations over one with more of them.
+
+Historically, it was designed to help with `ExpressibleByXYZLiteral` types.
+The type checker strongly prefers to give literals their default type
+(e.g. `Int` for `ExpressibleByIntegerLiteral`,
+`String` for `ExpressibleByStringLiteral`, etc.).
+If an API should prefer some other type, but accept the default too,
+marking the declaration taking the default type with `@_disfavoredOverload`
+gives the desired behavior:
+
+```swift
+extension LocalizedStringKey: ExpressibleByStringLiteral { ... }
+
+extension Text {
+  // We want `Text("foo")` to use this initializer:
+  init(_ key: LocalizedStringKey) { ... }
+
+  // But without @_disfavoredOverload, it would use this one instead,
+  // because that lets it give the literal its default type:
+  @_disfavoredOverload init<S: StringProtocol>(_ str: S) { ... }
+}
+```
+
+`@_disfavoredOverload` is a stop-gap design; eventually the type-checker
+behavior should be fixed and the attribute should be removed/made into a no-op.
+
+Its main use case is to work around clearly undesirable overload resolution,
+not to arbitrarily manipulate overload priority.
+
+## `@_dynamicReplacement(for:)`
+
+Marks a function as the dynamic replacement for another `dynamic` function.
+This is similar to method swizzling in other languages such as Objective-C,
+except that the replacement happens at program start (or loading a shared
+library), instead of at an arbitrary point in time.
+
+For more details, see [Dynamic method replacement](https://forums.swift.org/t/dynamic-method-replacement/16619).
+
+## `@_distributedActorIndependent`
+
+## `@_effects`
+
+## `@_exported`
+
+Use to mark an imported module to re-export all its declarations.
+This attribute is most commonly used by overlays.
+
+```swift
+// module M
+func f() {}
+
+// module N
+@_exported import M
+
+// module P
+import N
+func g() {
+  N.f() // OK
+}
+```
+
+## `@_fixed_layout`
+
+## `@_hasInitialValue`
+
+## `@_hasMissingDesignatedInitializers`
```

Comment:
Proposing a definition here:

An attribute that indicates that there may be designated initializers that are not printed in the swiftinterface file for a particular class. This attribute is needed for the initializer model to maintain correctness when library evolution is enabled. This is because a class may have non-public designated initializers, and Swift allows the inheritance of convenience initializers if and only if the subclass overrides (or has synthesized overrides) of every designated initializer in its superclass. To see the problem in action, consider

```
// Lib.swift
open class A {
  init(invisible: ()) {}

  public init(visible: ()) {}
  public convenience init(hi: ()) { self.init(invisible: ()) }
}

// Client.swift
class B : A {
  var x: String
  
  public override init(visible: ()) {
    self.x = "Garbage"
    super.init(visible: ())
  }
}
```

In this case, if B were allowed to inherit the convenience initializer `A.init(invisible:)` then an instance created via `B(hi: ())` would fail to initialize `B.x` resulting in a memory safety hole. What's worse is there is no way to close this safety hole because the user cannot override the invisible designated initializer because they lack sufficient visibility.